libbindgen/src/codegen/mod.rs

```diff
@@ -650,7 +650,13 @@ impl CodeGenerator for CompInfo {
 
         let mut template_args_used =
             vec![false; applicable_template_args.len()];
-        let canonical_name = item.canonical_name(ctx);
+
+        let mut canonical_name = item.canonical_name(ctx);
+        let times_seen = result.overload_number(&canonical_name);
```

Comment:
For function overloads this is fine, but for structs it's not, since we can't disambiguate them when we refer to them. I think this logic belongs to `canonical_name`, and indeed I think it worked at some point (see the namespacing and `parent_canonical_name` logic), it just needs to be tuned so it works again.